Add TCPC driver for FUSB307

recalci opened the pull request:
This PR adds the TCPC driver for the FUSB307 and includes several related fixes.

recalci committed:
usbc: Enable VBUS measurement before VBUS check level

Enable VBUS measurement before checking the VBUS level and
disable it when unattached.

Signed-off-by: Jianxiong Gu <jianxiong.gu@outlook.com>

recalci committed:
drivers: tcpc: Add TCPC driver for FUSB307

Add support for FUSB307.

Signed-off-by: Jianxiong Gu <jianxiong.gu@outlook.com>

recalci committed:
drivers: tcpc: ps8xxx: Update to use generic TCPCI function

Update set_snk_ctrl, set_src_ctrl, and set_low_power_mode
to use the generic TCPCI function.

Signed-off-by: Jianxiong Gu <jianxiong.gu@outlook.com>

recalci committed:
usbc: tcpc: Add generic TCPCI related functions

Add generic functions that will be common to TCPCI compliant drivers.

Signed-off-by: Jianxiong Gu <jianxiong.gu@outlook.com>

recalci committed:
tests: drivers: Add FUSB307 driver

Add a build test to verify fusb307 driver builds correctly.

Signed-off-by: Jianxiong Gu <jianxiong.gu@outlook.com>

keith-zephyr reviewed (commented):
The first commits on this PR are bug fixes and could be merged into the 4.1 release if you split them out into a separate PR.

recalci commented on drivers/usb_c/tcpc/fusb307.c:
```diff
@@ -0,0 +1,760 @@
+		return ret;
+	}
+
+	tcpci_tcpm_get_cc(&cfg->bus, &cc1, &cc2);
+	if (cc1) {
+		tcpci_update_reg8(&cfg->bus, TCPC_REG_ROLE_CTRL, TCPC_REG_ROLE_CTRL_CC2_MASK,
+				  TCPC_REG_ROLE_CTRL_SET(0, 0, 0, TC_CC_OPEN));
+	} else if (cc2) {
+		tcpci_update_reg8(&cfg->bus, TCPC_REG_ROLE_CTRL, TCPC_REG_ROLE_CTRL_CC1_MASK,
+				  TCPC_REG_ROLE_CTRL_SET(0, 0, TC_CC_OPEN, 0));
+	} else {
+		if (data->power_role == TC_ROLE_SINK) {
+			tcpci_tcpm_set_cc(&cfg->bus, TC_CC_RD);
+		} else {
+			tcpci_tcpm_set_cc(&cfg->bus, TC_CC_RP);
+		}
+	}
```
Based on testing, FUSB307 won't respond with GoodCRC on CC2 unless CC1 is set to OPEN. [Chrome EC](https://github.com/coreboot/chrome-ec/blob/ad41bf945927e186eaad768a9d51f45ada67b9b9/driver/tcpm/fusb307.c#L37) and ST's x-cube-usb-pd handle this the same way.

keith-zephyr replied:
Thanks for the explanation.  please add a comment describing this behavior.

recalci commented:
> The first commits on this PR are bug fixes and could be merged into the 4.1 release if you split them out into a separate PR.

Thank you for your review. I've created a new PR [#86143](https://github.com/zephyrproject-rtos/zephyr/pull/86143)

keith-zephyr commented on drivers/usb_c/tcpc/rt1715.c:
```diff
@@ -427,8 +427,9 @@ static int rt1715_tcpc_mask_status_register(const struct device *dev, enum tcpc_
 static int rt1715_tcpc_set_drp_toggle(const struct device *dev, bool enable)
 {
 	const struct rt1715_cfg *cfg = dev->config;
+	uint8_t pd_int_rev = 0x10;
```
You should read this revision from the chip register directly during the init function and save it in the `struct rt1715_data` of the driver.

keith-zephyr commented on drivers/usb_c/tcpc/tcpci.c:
```diff
@@ -598,10 +598,36 @@ int tcpci_tcpm_set_roles(const struct i2c_dt_spec *bus, enum pd_rev_type pd_rev,
+	 * For TCPCI Rev 2.0, unless the TCPM sets TCPC_REG_TCPC_CTRL_EN_LOOK4CONNECTION_ALERT,
+	 * TCPC by default masks Alert assertion when CC_STATUS_LOOK4CONNECTION changes state.
+	 */
+	if (pd_int_rev >= 0x20) {
```
Create macro for revision 0x20.

keith-zephyr commented on drivers/usb_c/tcpc/fusb307.c:
```diff
@@ -0,0 +1,761 @@
+static int fusb307_tcpc_set_drp_toggle(const struct device *dev, bool enable)
+{
+	const struct fusb307_cfg *cfg = dev->config;
+	uint8_t pd_int_rev = 0x10;
```
Same as the rt1715 driver - read this from the chip register and store a value in the driver data.

recalci replied:
Done. I have updated both rt1715 and fusb307 as suggested.

recalci committed:
usbc: fix errors in TCPC_REG_PD_REV and TCPC_REG_PD_INT_REV macros

Corrected macro definitions.

Signed-off-by: Jianxiong Gu <jianxiong.gu@outlook.com>

recalci committed:
usbc: tcpc: Update set_drp_toggle function

Updated set_drp_toggle to handle differences in TCPCI revisions.
Added a macro for TCPCI revision and read it from the chip register
during initialization.

Signed-off-by: Jianxiong Gu <jianxiong.gu@outlook.com>

keith-zephyr commented on drivers/usb_c/tcpc/tcpci.c:
```diff
@@ -598,10 +598,37 @@ int tcpci_tcpm_set_roles(const struct i2c_dt_spec *bus, enum pd_rev_type pd_rev,
 }
 
-int tcpci_tcpm_set_drp_toggle(const struct i2c_dt_spec *bus, bool enable)
+int tcpci_tcpm_set_drp_toggle(const struct i2c_dt_spec *bus, enum pd_int_rev_type pd_int_rev,
```
I don't think using an `enum` for the TCPCi revision makes sense.  Create macros for `PD_INT_REV10` and `PD_INT_REV20` instead and pass a `uint8_t` type to this function.

recalci replied:
Done. Thank you for your patience

keith-zephyr commented on include/zephyr/drivers/usb_c/tcpci_priv.h:
```diff
@@ -41,6 +41,16 @@ struct tcpci_reg_dump_map {
+/**
+ * @brief USB-Port Controller Interface Specification Revision
+ */
+enum pd_int_rev_type {
```
Replace this enum with C macros.

recalci replied:
Done.

jfischer-no commented on drivers/usb_c/tcpc/tcpci.c:
```diff
@@ -682,3 +682,54 @@ int tcpci_tcpm_mask_status_register(const struct i2c_dt_spec *bus, enum tcpc_sta
+
+	ret = tcpci_tcpm_get_status_register(bus, TCPC_REG_POWER_STATUS, &reg);
+	if (ret == 0) {
+		*sinking = !!(reg & TCPC_REG_POWER_STATUS_SINKING_VBUS);
```
Should be just `*sinking = reg & TCPC_REG_POWER_STATUS_SINKING_VBUS;`, see C STD scalar value conversion to bool.

jfischer-no commented on drivers/usb_c/tcpc/tcpci.c:
```diff
@@ -682,3 +682,54 @@ int tcpci_tcpm_mask_status_register(const struct i2c_dt_spec *bus, enum tcpc_sta
+
+	ret = tcpci_tcpm_get_status_register(bus, TCPC_REG_POWER_STATUS, &reg);
+	if (ret == 0) {
+		*sourcing = !!(reg & TCPC_REG_POWER_STATUS_SOURCING_VBUS);
```
see above

jfischer-no commented on drivers/usb_c/tcpc/tcpci.c:
```diff
@@ -598,10 +598,36 @@ int tcpci_tcpm_set_roles(const struct i2c_dt_spec *bus, enum pd_rev_type pd_rev,
+		}
+	}
+
+	if (enable) {
+		/* Set Look4Connection command */
+		return tcpci_write_reg8(bus, TCPC_REG_COMMAND, TCPC_REG_COMMAND_LOOK4CONNECTION);
+	} else {
+		return 0;
+	}
```
```c
	if (enable) {
		/* Set Look4Connection command */
		return tcpci_write_reg8(bus, TCPC_REG_COMMAND, TCPC_REG_COMMAND_LOOK4CONNECTION);
	}

	return 0;
```
IIRC, there should be checkpatch warning for this kind of constructs.

jfischer-no commented on drivers/usb_c/tcpc/fusb307.c:
```diff
@@ -0,0 +1,770 @@
+#define DT_DRV_COMPAT onnn_fusb307_tcpc
+LOG_MODULE_REGISTER(tcpc_fusb307, CONFIG_USBC_LOG_LEVEL);
+
+/** Data structure for device instances */
```
Comment style must be /* */, it is documented.
Also, there is no need to comment obvious things like `/* Data structure for device instances */`, `Delayable work`, `Boolean valu`, `State of CC1`...

jfischer-no commented on drivers/usb_c/tcpc/fusb307.c:
```diff
@@ -0,0 +1,770 @@
+		VCONN_DISC_GPIO(node),                                                             \
+	}
+
+#define FUSB307_DRIVER_INIT(inst)                                                                  \
```
`#define FUSB307_INSTANCE_DEFINE(inst)`

jfischer-no commented on drivers/usb_c/tcpc/fusb307.c:
```diff
@@ -0,0 +1,770 @@
+	static struct fusb307_data drv_data_fusb307##inst =                                        \
+		FUSB307_DRIVER_DATA_INIT(DT_DRV_INST(inst));                                       \
+	static struct fusb307_cfg drv_cfg_fusb307##inst =                                          \
+		FUSB307_DRIVER_CFG_INIT(DT_DRV_INST(inst));                                        \
```
Just
```c

	static struct fusb307_cfg fusb307_cfg_##inst = {
		.bus = I2C_DT_SPEC_GET(node),                                                      \
		.alert_gpio = GPIO_DT_SPEC_GET(node, irq_gpios),                                   \
		.transmit_retries = DT_PROP(node, transmit_retries),                               \
		VCONN_DISC_GPIO(node),                                                             \
	};
```
and the same for the data

jfischer-no commented on drivers/usb_c/tcpc/fusb307.h:
```diff
@@ -0,0 +1,61 @@
+#define FUSB307_REG_ALERT_VD_MASK      0xB4
+
+/* Sampling period of VBUS Measurement, at least greater than 3ms */
+#define FUSB307_tVBUSsample 4 /* ms */
```
`#define FUSB307_T_VBUS_SAMPLE 4`

recalci commented:
@jfischer-no  thanks for the review.
Requested changes implemented

recalci commented:
Sorry，I closed this PR by mistake. Reopen

jfischer-no commented on dts/bindings/tcpc/onnn,fusb307.yaml:
```diff
@@ -0,0 +1,64 @@
+
+    Example:
+
+    ports {
+        #address-cells = <1>;
+        #size-cells = <0>;
+        port0: usbc-port@0 {
+          compatible = "usb-c-connector";
+          reg = <0>;
+          tcpc = <&fusb307_tcpc0>;
+          vbus = <&fusb307_vbus0>;
+          power-role = "sink";
+          sink-pdos = <PDO_FIXED(5000, 100, 0)>;
+        };
+    };
```
These lines do not look properly indented, please fix them.

jfischer-no reviewed (approved):
LGTM

craigpeacock commented:
What is the status on this one? @sambhurst
